#167510280 Create database schema

Nennyfills opened the pull request:
#### What does this PR do?
This pull request create the database schema
#### Description of Task to be completed
- Created migration folder
- Created migration files for all models
- Created model files for all database schema
- Added migration script  to the package.json file
- Added seed script to the package.json file

#### How should this be manually tested?
- git clone this branch
- run `npm db:migrate` to migrate the database schema
- run `npm db:undo` to revert the database schema

#### Any Background Context You want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#167510280](https://www.pivotaltracker.com/story/show/167510280)

#### Screenshots (if appropriate)
N/A

#### Questions:
N/A

Merkll commented on src/db/migrations/20190728220728-create-article.js:
```diff
@@ -0,0 +1,71 @@
+          allowNull: false,
+          type: Sequelize.STRING
+        },
+        timeToRead: {
```
We are no longer implementing time taken to read

successgilli reviewed (approved):
I will really appreciate if the model folder was moved to the root of the src folder just outside the db folder.
Great job @Nennyfills.

iverenshaguy commented on .sequelizerc:
```diff
@@ -1,6 +1,6 @@
-  'seeders-path': './config/seeders',
-  'migrations-path': './config/migrations'
+  'config': './src/db/sequelize.js',
+  'models-path': './src/db/models',
```
Kindly move this back to the src folder.
```suggestion
  'models-path': './src/models',
```

Nennyfills replied:
done.

iverenshaguy commented on .sequelizerc:
```diff
@@ -1,6 +1,6 @@
-  'models-path': './src/models',
-  'seeders-path': './config/seeders',
-  'migrations-path': './config/migrations'
+  'config': './src/db/sequelize.js',
```
Please move this into the src folder
```suggestion
  'config': './src/sequelize.js',
```

Nennyfills replied:
done

iverenshaguy commented on .sequelizerc:
```diff
@@ -1,6 +1,6 @@
-  'migrations-path': './config/migrations'
+  'config': './src/db/sequelize.js',
+  'models-path': './src/db/models',
+  'seeders-path': './src/db/seeders',
```
Remove the seeders and migrations from the src folder. Move it to the config folder outside the source folder since we won't need to transpile it i.e. it is only needed for setup.
```suggestion
  'seeders-path': './config/db/seeders',
  'migrations-path': './config/db/migrations'
```

Nennyfills replied:
done.

iverenshaguy commented on package.json:
```diff
@@ -6,43 +6,33 @@
+    "test": "mocha --require @babel/register --require regenerator-runtime/runtime tests/test.js --exit",
+    "test:coverage": "nyc npm run test",
+    "sequelize-cli:es6": "babel-node ./node_modules/.bin/sequelize $*",
+    "db:rollback": "npm run db:undo && npm run db:migrate && npm run db:seed",
```
Kindly arrange the scripts alphabetically

Nennyfills replied:
done

iverenshaguy commented on package.json:
```diff
@@ -6,43 +6,33 @@
-    "request": "^2.87.0",
-    "swagger-ui-express": "^4.0.7",
-    "underscore": "^1.9.1"
+    "sequelize-cli": "^5.5.0",
```
Move `sequelize-cli` back to dev dependencies

Nennyfills replied:
done

iverenshaguy commented on src/db/migrations/20190728220710-create-user.js:
```diff
@@ -0,0 +1,63 @@
+
+const debug = Debug('dev');
+
+module.exports = {
```
Change to es6 i.e. 
```suggestion
export default {
```
or doesn't this work? Please do the same thing for all migration and seeder files

Nennyfills replied:
The module.exports was generated from sequelize by default, will have to return back the sequelize.js to the db folder for the export default to work on the migration files.

iverenshaguy commented on src/db/migrations/20190728221040-create-downVote.js:
```diff
@@ -0,0 +1,49 @@
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    try {
+      await queryInterface.createTable('downVote', {
```
rename file to `...-create-downvote` and then rename the table to `Downvotes`
```suggestion
      await queryInterface.createTable('Downvotes', {
```

Nennyfills replied:
done

iverenshaguy commented on src/db/migrations/20190728221040-create-downVote.js:
```diff
@@ -0,0 +1,49 @@
+  },
+  down: async (queryInterface) => {
+    try {
+      await queryInterface.dropTable('downVote');
```
Same as above

iverenshaguy commented on src/db/migrations/20190728221104-create-upVote.js:
```diff
@@ -0,0 +1,49 @@
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    try {
+      await queryInterface.createTable('UpVote', {
```
Rename the file to `...-create-upvote` and rename the table to `Upvotes`

Nennyfills replied:
done

iverenshaguy commented on src/db/migrations/20190728221239-create-notification.js:
```diff
@@ -0,0 +1,51 @@
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    try {
+      await queryInterface.createTable('Notification', {
```
```suggestion
      await queryInterface.createTable('Notifications', {
```
Please do the same thing to `down`

Nennyfills replied:
done

iverenshaguy commented on src/db/migrations/20190728221239-create-notification.js:
```diff
@@ -0,0 +1,51 @@
+  },
+  down: async (queryInterface) => {
+    try {
+      await queryInterface.dropTable('AuNotificationthor');
```
??? I think you should actually run and undo the migrations to catch errors like this one
```suggestion
      await queryInterface.dropTable('Notifications');
```

Nennyfills replied:
Done.

iverenshaguy commented on src/db/migrations/20190728221340-create-readStat.js:
```diff
@@ -0,0 +1,44 @@
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    try {
+      await queryInterface.createTable('ReadStat', {
```
Rename the file to `...-create-readstat` and rename the table to ReadStats
```suggestion
      await queryInterface.createTable('ReadStats', {
```

Nennyfills replied:
Done.

iverenshaguy commented on src/db/migrations/20190728221410-create-report.js:
```diff
@@ -0,0 +1,48 @@
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    try {
+      await queryInterface.createTable('Report', {
```
```suggestion
      await queryInterface.createTable('Reports', {
```

Nennyfills replied:
Done.

iverenshaguy commented on src/db/migrations/20190728221340-create-readStat.js:
```diff
@@ -0,0 +1,44 @@
+  },
+  down: async (queryInterface) => {
+    try {
+      await queryInterface.dropTable('ReadStat');
```
```suggestion
      await queryInterface.dropTable('ReadStats');
```

Nennyfills replied:
Done.

iverenshaguy commented on src/db/migrations/20190728221410-create-report.js:
```diff
@@ -0,0 +1,48 @@
+  },
+  down: async (queryInterface) => {
+    try {
+      await queryInterface.dropTable('Report');
```
```suggestion
      await queryInterface.dropTable('Reports');
```

Nennyfills replied:
Done

iverenshaguy commented on src/db/migrations/20190728221432-create-roles.js:
```diff
@@ -0,0 +1,42 @@
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    try {
+      await queryInterface.createTable('Article', {
```
The migration file name is `create-roles` while the table name is `Article` which should be `Articles` by the way. Are we missing something?

iverenshaguy commented on src/db/migrations/20190729144905-create-blacklist.js:
```diff
@@ -0,0 +1,43 @@
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    try {
+      await queryInterface.createTable('BooklistedToken', {
```
Filename is `...create-blacklist` and table name is `BooklistedToken`? I'm confused.

Nennyfills replied:
changed it to blacklistedTokens. Done.

iverenshaguy commented on src/db/models/BooklistedToken.js:
```diff
@@ -0,0 +1,14 @@
+export default (sequelize, DataTypes) => {
+  const BooklistedToken = sequelize.define('BooklistedToken', {
```
```suggestion
  const BooklistedToken = sequelize.define('BooklistedTokens', {
```
I think this should be `Blacklisted` not `Booklisted`

Nennyfills replied:
Yes, done

iverenshaguy commented on src/db/migrations/20190728220750-create-articleCategory.js:
```diff
@@ -0,0 +1,44 @@
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    try {
+      await queryInterface.createTable('ArticleCategory', {
```
```suggestion
      await queryInterface.createTable('ArticleCategories', {
```

Nennyfills replied:
Done.

iverenshaguy commented on src/db/models/articleCategory.js:
```diff
@@ -0,0 +1,21 @@
+    {
+      authorId: DataTypes.UUID,
+      categoryId: DataTypes.UUID
+    }, {}
```
```suggestion
    },
    {}
```
Please be consistent in your indentations. Either us this style or use the one you used for the Author Model

Nennyfills replied:
Okay. Done!

iverenshaguy commented on src/db/models/articleCategory.js:
```diff
@@ -0,0 +1,21 @@
+      onDelete: 'CASCADE'
+    });
+  };
+  return ArticleCategory;
```
```suggestion

  return ArticleCategory;
```

Nennyfills replied:
Done.

iverenshaguy commented on src/db/models/author.js:
```diff
@@ -0,0 +1,18 @@
+    userId: DataTypes.UUID,
+    noOfPublication: DataTypes.BIGINT
+  }, {});
+  Author.associate = (models) => {
```
```suggestion

  Author.associate = (models) => {
```

Nennyfills replied:
Done.

iverenshaguy commented on src/db/models/bookmark.js:
```diff
@@ -0,0 +1,19 @@
+      onDelete: 'CASCADE'
+    });
+  };
+  return Bookmark;
```
```suggestion

  return Bookmark;
```

Nennyfills replied:
Done

iverenshaguy commented on src/db/models/category.js:
```diff
@@ -0,0 +1,16 @@
+      description: DataTypes.STRING
+    }, {}
+  );
+  Category.associate = (models) => {
```
```suggestion

  Category.associate = (models) => {
```

iverenshaguy commented on src/db/models/category.js:
```diff
@@ -0,0 +1,16 @@
+      onDelete: 'CASCADE'
+    });
+  };
+  return Category;
```
```suggestion

  return Category;
```

Nennyfills replied:
Done

iverenshaguy commented on src/db/models/comments.js:
```diff
@@ -0,0 +1,25 @@
+      onDelete: 'CASCADE'
+    });
+  };
+  return Comments;
```
```suggestion

  return Comments;
```

iverenshaguy commented on src/db/models/downVote.js:
```diff
@@ -0,0 +1,22 @@
+export default (sequelize, DataTypes) => {
+  const downVote = sequelize.define('downVote', {
```
```suggestion
  const Downvote = sequelize.define('Downvote', {
```

iverenshaguy replied:
Update the whole file as needed

iverenshaguy commented on src/db/models/downVote.js:
```diff
@@ -0,0 +1,22 @@
+      onDelete: 'CASCADE'
+    });
+  };
+  return downVote;
```
```suggestion

  return downVote;
```

iverenshaguy commented on src/db/models/followers.js:
```diff
@@ -0,0 +1,18 @@
+      onDelete: 'CASCADE'
+    });
+  };
+  return Followers;
```
```suggestion

  return Followers;
```

iverenshaguy commented on src/db/models/notification.js:
```diff
@@ -0,0 +1,15 @@
+    payload: DataTypes.JSON,
+    notified: DataTypes.BOOLEAN,
+  }, {});
+  Notification.associate = (models) => {
```
```suggestion

  Notification.associate = (models) => {
```

iverenshaguy commented on src/db/models/notification.js:
```diff
@@ -0,0 +1,15 @@
+      onDelete: 'CASCADE'
+    });
+  };
+  return Notification;
```
```suggestion

  return Notification;
```

iverenshaguy commented on src/db/models/permissions.js:
```diff
@@ -0,0 +1,19 @@
+    code: DataTypes.BIGINT,
+    roleId: DataTypes.UUID
+  }, {});
+  Permissions.associate = (models) => {
```
```suggestion

  Permissions.associate = (models) => {
```

iverenshaguy commented on src/db/models/permissions.js:
```diff
@@ -0,0 +1,19 @@
+      onDelete: 'CASCADE'
+    });
+  };
+  return Permissions;
```
```suggestion

  return Permissions;
```

iverenshaguy commented on src/db/models/readStat.js:
```diff
@@ -0,0 +1,18 @@
+      onDelete: 'CASCADE'
+    });
+  };
+  return ReadStat;
```
```suggestion

  return ReadStat;
```

iverenshaguy commented on src/db/models/report.js:
```diff
@@ -0,0 +1,18 @@
+      foreignKey: 'articleId',
+    });
+  };
+  return Report;
```
```suggestion

  return Report;
```

iverenshaguy commented on src/db/models/roles.js:
```diff
@@ -0,0 +1,13 @@
+    roleName: DataTypes.STRING,
+    description: DataTypes.STRING
+  }, {});
+  Role.associate = (models) => {
```
```suggestion

  Role.associate = (models) => {
```

iverenshaguy commented on src/db/models/roles.js:
```diff
@@ -0,0 +1,13 @@
+      onDelete: 'CASCADE'
+    });
+  };
+  return Role;
```
```suggestion

  return Role;
```

iverenshaguy commented on src/db/models/upVote.js:
```diff
@@ -0,0 +1,23 @@
+      onDelete: 'CASCADE'
+    });
+  };
+  return UpVote;
```
```suggestion

  return UpVote;
```

Nennyfills replied:
Done

iverenshaguy commented on src/db/models/user.js:
```diff
@@ -0,0 +1,54 @@
+    roleId: DataTypes.UUID
+
+  }, {});
+  User.associate = (models) => {
```
```suggestion

  User.associate = (models) => {
```

Nennyfills replied:
Done.

iverenshaguy commented on src/db/models/user.js:
```diff
@@ -0,0 +1,54 @@
+    password: DataTypes.STRING,
+    avatarUrl: DataTypes.STRING,
+    roleId: DataTypes.UUID
+
```
```suggestion
```

Nennyfills replied:
Done.

iverenshaguy commented on src/db/models/user.js:
```diff
@@ -0,0 +1,54 @@
+      foreignKey: 'userId',
+      onDelete: 'CASCADE'
+    });
+    User.hasMany(models.Report, {
```
Leave new lines between all the associations
```suggestion

    User.hasMany(models.Report, {
```

Nennyfills replied:
Done.

iverenshaguy commented on src/db/models/user.js:
```diff
@@ -0,0 +1,54 @@
+      onDelete: 'CASCADE'
+    });
+  };
+  return User;
```
```suggestion

  return User;
```

Nennyfills replied:
Done.

iverenshaguy reviewed (changes requested):
Kindly review my comments.

Other things to note
- Always name migration files in lowercase
- Always create pluralized migration tables
- Always leave a new line before return statements
- Always test migrations thoroughly after creation (up and down)
- Pay more attention to details
- Your PR title should be in sentence case

iverenshaguy commented on .nycrc:
```diff
@@ -3,7 +3,7 @@
     "regenerator-runtime/runtime"
     ],
-  "reporter": [
+  "Reportser": [
```
Reportser????

Nennyfills committed:
feat(create): create database schema

- create migration folder
- create migration files for all models
- create model files for all database schema
- add migration script  to the package.json file
- add seed script to the package.json file

[Starts #167510280]

iverenshaguy committed:
feat(database): update implementation

- rename tables

[Finishes #167510280]
